docs/research/format_to_parts.md

```diff
@@ -0,0 +1,190 @@
+Format to Parts in ICU4X
+========================
+
+ECMA-402 specifies *[formatToParts](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/formatToParts#Description)*, a variant of the *format* function that returns the string split into segments according to what the segment represents.  For example:
+
+```javascript
+new Intl.DateTimeFormat("fr").format(new Date())
+// "19/11/2020"
+
+new Intl.DateTimeFormat("fr").formatToParts(new Date())
+/* [
+  {type: "day", value: "19"},
+  {type: "literal", value: "/"},
+  {type: "month", value: "11"},
+  {type: "literal", value: "/"},
+  {type: "year", value: "2020"},
+] */
+
+new Intl.DateTimeFormat("fr", { dateStyle: "medium" }).formatRange(new Date(2020, 0, 1), new Date(2020, 2, 3))
+// "1 janv. – 3 mars 2020"
+
+new Intl.DateTimeFormat("fr", { dateStyle: "medium" }).formatRangeToParts(new Date(2020, 0, 1), new Date(2020, 2, 3))
+/* [
+  {type: "day", value: "1", source: "startRange"},
+  {type: "literal", value: " ", source: "startRange"},
+  {type: "month", value: "janv.", source: "startRange"},
+  {type: "literal", value: " – ", source: "shared"},
+  {type: "day", value: "3", source: "endRange"},
+  {type: "literal", value: " ", source: "endRange"},
+  {type: "month", value: "mars", source: "endRange"},
+  {type: "literal", value: " ", source: "shared"},
+  {type: "year", value: "2020", source: "shared"},
+] */
+```
+
+The goal of this document is to discuss how we will deal with formatToParts in ICU4X, both with regard to the API model and the data model (implementation).
+
+## API Model
+
+There are two prevailing models in formatting to parts: linear attributes and nested fields.
+
+### Linear Attributes
+
+ECMA-402 uses the linear attributes model: each substring is mapped to exactly one bag of attributes.  Expressed in terms of starts and ends of ranges:
+
+```
+1 janv. – 3 mars 2020
+
+[) day, startRange
+ [) literal, startRange
+  [    ) month, startRange
+       [  ) literal, shared
+          [) day, endRange
+           [) literal, endRange
+            [   ) month, endRange
+                [) literal, shared
+                 [   ) year, Shared
+```
+
+### Nested Fields
+
+In the nested fields model, there is a set of ranges, and a particular character could live in more than one range.  Here is how the above date interval format would look with nested fields:
+
+```
+1 janv. – 3 mars 2020
+
+[) day
+ [) literal
+  [    ) month
+[      ) startRange
+       [  ) literal
+       [  ) shared
+          [) day
+           [) literal
+            [   ) month
+          [     ) endRange
+                [) literal
+                 [   ) year
+                [    ) shared
+```
+
+The linear attributes model is simpler, but the nested fields model is more flexible.
```

Comment:
What requirements are we trying to fulfil here?  Is the requirement just to reimplement a spec, or are there other user driven things that we want to do?  Perhaps list them here, as it is hard to compare alternatives without a reference as to what we're after.

Reply:
For example, is the ability to iterate over the output parts at different levels of granularity required, or not?

Do intervals at most nest, or can they also overlap?

Reply:
Actually, I really don't have a good answer to this question. By my understanding, the main reason to use formatToParts is to apply different styles to different parts of a string. Linear attributes cover this use case. I am hoping @mihnita and @zbraniecki can help clarify what they see as the key requirements that can inform the design decision here.

Reply:
Like Mihai said, in ECMA-402 motivation was purely to allow for styling of parts - "January <b>4th</b> 2020" or "10:00<b>am</b>".

But I think that this may not be enough for our purpose. ECMA-402 is optimizing for the smallest subset necessary to supply the needs.
We should aim to build a superset of that. And nesting of levels of parts may be something important here.

An example taxonomy may be:

| char | Lvl1 | Lvl2 |
| --- | ----------- |--|
| M | Weekday(start) | Weekday(start) |
| o | Weekday(middle) | Weekday(middle) |
| n | Weekday(end) | Weekday(end) |
| : | Literal(single) | Literal(single) |
| 1 | Hour(start) | Time(start) |
| 0 | Hour(end) | Time(middle) |
| : | Literal(single) | Time(middle) |
| 3 | Minute(start) | Time(middle) |
| 8 | Minute(end) | Time(middle) |
| a | DayPeriod(start) | Time(middle) |
| m | DayPeriod(end) | Time(end) |

In such scenario you may want to iterate over it on different "levels". I'm not sure if my example is clear, but I hope you can derive the meaning :)

I can see an argument that it may be hard, fragile or even impossible to try to capture the "Time" part just from iterating over "Lvl1", so we'd need a nested levels to identify parts you want to capture and in cases like ranges there may be more levels.